packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -41,97 +43,149 @@ const TextStyle _kToolbarButtonFontStyle = TextStyle(
   color: CupertinoColors.white,
 );
 
+/// The direction of the triangle attached to the toolbar.
+///
+/// Defaults to showing the triangle downwards if sufficient space is available
+/// to show the toolbar above the text field. Otherwise, the toolbar will
+/// appear below the text field and the triangle's direction will be [up].
+enum _ArrowDirection { up, down }
+
 /// Paints a triangle below the toolbar.
 class _TextSelectionToolbarNotchPainter extends CustomPainter {
+  const _TextSelectionToolbarNotchPainter(
+    this.arrowDirection,
+    this.triangleOffsetX,
+  ) : assert (arrowDirection != null),
+      assert (triangleOffsetX != null);
+
+  final _ArrowDirection arrowDirection;
+  final double triangleOffsetX;
+
   @override
   void paint(Canvas canvas, Size size) {
     final Paint paint = Paint()
-        ..color = _kToolbarBackgroundColor
-        ..style = PaintingStyle.fill;
+      ..color = _kToolbarBackgroundColor
+      ..style = PaintingStyle.fill;
+    final double triangleBottomY = (arrowDirection == _ArrowDirection.down)
+      ? 0.0
+      : _kToolbarTriangleSize.height;
     final Path triangle = Path()
-        ..lineTo(_kToolbarTriangleSize.width / 2, 0.0)
-        ..lineTo(0.0, _kToolbarTriangleSize.height)
-        ..lineTo(-(_kToolbarTriangleSize.width / 2), 0.0)
-        ..close();
+      ..moveTo(triangleOffsetX, 0)
+      ..lineTo(triangleOffsetX + _kToolbarTriangleSize.width / 2, triangleBottomY)
+      ..lineTo(triangleOffsetX + 0.0, _kToolbarTriangleSize.height)
+      ..lineTo(triangleOffsetX -(_kToolbarTriangleSize.width / 2), triangleBottomY)
+      ..close();
     canvas.drawPath(triangle, paint);
   }
 
   @override
-  bool shouldRepaint(_TextSelectionToolbarNotchPainter oldPainter) => false;
+  bool shouldRepaint(_TextSelectionToolbarNotchPainter oldPainter) {
+    return triangleOffsetX != oldPainter.triangleOffsetX;
+  }
 }
 
 /// Manages a copy/paste text selection toolbar.
-class _TextSelectionToolbar extends StatelessWidget {
+class _TextSelectionToolbar extends StatefulWidget {
   const _TextSelectionToolbar({
     Key key,
     this.handleCut,
     this.handleCopy,
     this.handlePaste,
     this.handleSelectAll,
+    this.arrowDirection,
   }) : super(key: key);
 
   final VoidCallback handleCut;
   final VoidCallback handleCopy;
   final VoidCallback handlePaste;
   final VoidCallback handleSelectAll;
+  final _ArrowDirection arrowDirection;
+
+  @override
+  _TextSelectionToolbarState createState() => _TextSelectionToolbarState();
+}
+
+class _TextSelectionToolbarState extends State<_TextSelectionToolbar> {
+  double _triangleOffsetX = double.nan;
```

Comment:
Could this be provided in the constructor instead of starting as nan and needing to call `updateTriangleOffsetX`?